ql/pricingengines/vanilla/coshestonengine.cpp

```diff
@@ -0,0 +1,275 @@
+        StringReplace[ a, {"Sqrt" -> "std::sqrt", "Power(E," -> "std::exp(",
+                           "Power" -> "std::pow"}]}, aa], {0, Infinity}]] &;
+
+    c[n_] := cpp@CForm[FullSimplify[Derivative[n][e][0],
+         kappa > 0 && theta > 0 && v0 > 0 && sigma > 0 &&
+         rho [Element] {-1, 1} && i^2 == -1]]
+    */
+
+   Real COSHestonEngine::c1(Time t) const {
+       return (-theta_ + std::exp(kappa_*t)
+           *(2*kappa_*muT(t) + theta_ - kappa_*t*theta_ -
+               v0_) + v0_)/(2*std::exp(kappa_*t)*kappa_);
+   }
+
+   Real COSHestonEngine::c2(Time t) const {
+       const Real sigma2 = sigma_*sigma_;
+       const Real kappa2 = kappa_*kappa_;
+       const Real kappa3 = kappa2*kappa_;
+
+       return (sigma2*(theta_ - 2*v0_) +
+           std::exp(2*kappa_*t)*(8*kappa3*t*theta_ -
+           8*kappa2*(theta_ + rho_*sigma_*t*theta_ - v0_) +
+           sigma2*(-5*theta_ + 2*v0_) + 2*kappa_*sigma_*(8*rho_*theta_ +
+           sigma_*t*theta_ - 4*rho_*v0_)) +
+           4*std::exp(kappa_*t)*(sigma2*theta_ -
+           2*kappa2*(-1 + rho_*sigma_*t)*(theta_ - v0_) +
+           kappa_*sigma_*(sigma_*t*(theta_ - v0_) + 2*rho_*(-2*theta_ +
+           v0_))))/(8.*std::exp(2*kappa_*t)*kappa3);
+   }
+
+   Real COSHestonEngine::c3(Time t) const {
+       const Real sigma2 = sigma_*sigma_;
+       const Real sigma3 = sigma2*sigma_;
+       const Real kappa2 = kappa_*kappa_;
+       const Real kappa3 = kappa2*kappa_;
+       const Real kappa4 = kappa3*kappa_;
+       const Real rho2   = rho_*rho_;
+
+       return
+         -(sigma_*(sigma3*(theta_ - 3*v0_) +
+           std::exp(3*kappa_*t)*(2*(-11*sigma3 -
+           24*kappa4*rho_*t + 3*kappa_*sigma2*(20*rho_ +
+           sigma_*t) - 6*kappa2*sigma_*(5 + 3*rho_*(4*rho_ + sigma_*t)) +
+           12*kappa3*(sigma_*t + 2*rho_*(2 + rho_*sigma_*t)))*theta_ -
+           6*(2*kappa_*rho_ - sigma_)*(4*kappa2 - 4*kappa_*rho_*sigma_ +
+           sigma2)*v0_) + 6*std::exp(kappa_*t)*sigma_*(-2*kappa2*(-1 +
+           rho_*sigma_*t)*(theta_ - 2*v0_) + sigma2*(theta_ - v0_) +
+           kappa_*sigma_*(-4*rho_*theta_ + sigma_*t*theta_ + 6*rho_*v0_ - 2*sigma_*t*v0_)) +
+           3*std::exp(2*kappa_*t)*(2*kappa_*sigma2*(-16*rho_*theta_ +
+           sigma_*t*(3*theta_ - v0_)) + 8*kappa4*rho_*t*(-2 +
+           rho_*sigma_*t)*(theta_ - v0_) + sigma3*(5*theta_ + v0_) +
+           8*kappa3*(-(rho_*(4 + sigma2*t*t)*theta_) + 2*sigma_*t*(theta_ - v0_) +
+           2*rho2*sigma_*t*(2*theta_ - v0_) + rho_*(2 +
+           sigma2*t*t)*v0_) + 2*kappa2*sigma_*((8
+           + 24*rho2 - 16*rho_*sigma_*t + sigma2*t*t)*theta_ - (8*rho2 -
+           8*rho_*sigma_*t + sigma2*t*t)*v0_))))/(16.*std::exp(3*kappa_*t)*
+           kappa_*kappa4);
+   }
+
+   Real COSHestonEngine::c4(Time t) const {
+       const Real sigma2 = sigma_*sigma_;
+       const Real sigma3 = sigma2*sigma_;
+       const Real sigma4 = sigma2*sigma2;
+       const Real sigma5 = sigma2*sigma3;
```

Comment:
I'm getting a warning that this one is unused - is this expected, or is there an error in the formula below?